src/main/java/com/google/sps/servlets/ShoppingCartServlet.java

```diff
@@ -0,0 +1,160 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.JSONObject;
+import org.json.JSONException;
+
+/**
+ * Servlet that manipulates the list of books in the shopping cart.
+ * Provides GET, POST and DELETE methods.
+ */
+
+@WebServlet("/shopping-cart")
+public class ShoppingCartServlet extends HttpServlet {
+  
+  private FirebaseAuthentication firebaseAuth;
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void init() throws ServletException {
+    // initialize firebase app 
+    try {
+      setFirebaseAuth(new FirebaseAuthentication(FirebaseSingletonApp.getInstance()));
+    } catch (IOException e) {
+      System.out.println(e.getMessage());
+      // the exception must be catched because init() does not throw IOException()
+    }
+  }
+
+  public void setFirebaseAuth(FirebaseAuthentication firebaseAuth) {
+    this.firebaseAuth = firebaseAuth;
+  }
+
+  /**
+   * Returns the current list of books the user has in the cart.
+   */
+  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get current user.
+    String uid = null;
+    try {
+      // get the idToken from query string
+      uid = firebaseAuth.getUserIdFromIdToken(request.getParameter("idToken"));
+    } catch (FirebaseAuthException e) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, e.getMessage());
+      return;
+    }
+    
+    Query query = new Query("UserCart").setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    ArrayList<String> books = new ArrayList<>();
+    if (entity != null) {
+      // User has items added to cart.
+      books.addAll((ArrayList<String>) entity.getProperty("books"));
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(books);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Adds a new book to the current cart of the user.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get current user.
+    String uid = null;
+    JSONObject jsonObject;
+    try {
+      jsonObject = new JSONObject(request.getReader().lines().collect(Collectors.joining(System.lineSeparator())));
+    } catch (JSONException e){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+      return;
+    }
+    String bookId = jsonObject.get("bookId").toString();
+    try {
+      // get the idToken from hidden input field
```

Comment:
This should be `// Get the idToken from jsonObject`, right?

Reply:
Done